man/multiwan.5
.\" The MIT License (MIT)
.\" Copyright © 2013 Steve Buzonas <steve@fancyguy.com>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the “Software”), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
.\" THE SOFTWARE.
.TH MULTIWAN 5 2013-12-21 "FancyGuy Technologies" "Linux Programmer's Manual"
.SH NAME
multiwan \- the multi wan connection failover tool
.SH "SYNOPSIS"
.B /etc/multiwan.conf
.SH "DESCRIPTION"
.LP
This file is required in order for the script to function properly.
.LP
The different configuration options are:
.TP
\fBPING_TARGET\fP
IP address or domain name to ping.  The script relies on a service
to always be available.
.TP
\fBPING_TIMEOUT\fP
Ping timeout in seconds
.TP
\fBSUCCESS_COUNT\fP
Number of consecutive successful ping attempts before considering the link to be \fIup\fP.
.TP
\fBFAILURE_COUNT\fP
Number of consecutive failed ping attempts before considering the link to be \fIdown\fP.
.TP
\fBWAN_IF[n]\fP
The interface name for the n external network connection.
.TP
\fBWAN_GW[n]\fP
The gateway address to route traffic through the interface specified in \fIWAN_IF[n]\fP.
.TP
\fBWAN_TABLE[n]\fP
The table name of the route for \fIWAN_IF[n]\fP to be utilized by \fIiproute2\fP.
.TP
\fBWAN_WEIGHT[n]\fP
The route weight for the interface specified in \fIWAN_IF[n]\fP.
.TP
\fBLOG_FILE\fP
Log to file instead of syslog\fP.
.TP
\fBLOG_FACILITY\fP
Syslog facility, defaults to local6\fP.
.TP
\fBLOG_PRIORITY\fP
Syslog priority, defaults to notice\fP.
.SH "SEE ALSO"
multiwan(8), ip(8)


debian/changelog
multiwan (2.0) unstable; urgency=log

  * Unlimited wan connections. [\#1](https://github.com/honzahommer/multiwan/pull/1) ([honzahommer](https://github.com/honzahommer))
  * CI. [\#2](https://github.com/honzahommer/multiwan/pull/2) ([honzahommer](https://github.com/honzahommer))

 -- Honza Hommer <honza@hommer.cz>  Tue, 11 Feb 2020 17:01:44 +0100

multiwan (1.0) unstable; urgency=low

  * Initial Release.

 -- Steve Buzonas <steve@fancyguy.com>  Sun, 22 Dec 2013 22:33:13 -0500


multiwan.sh
#!/usr/bin/env bash

PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin

readonly __BASENAME="$(basename "$0" .sh)"

CONFIG="/etc/$__BASENAME.conf"

STARTUP=1
CHECK_INTERVAL=1
LOG_FACILITY=local6
LOG_PRIORITY=notice

declare -A WAN_IF
declare -A WAN_GW
declare -A WAN_TABLE
declare -A WAN_WEIGHT

declare -A WAN_IP # IP address of each WAN interface
declare -A LLS # Last link status. Defaults to down to force check of both on first run.
declare -A LPS # Last ping status.
declare -A CPS # Current ping status.
declare -A CLS # Change link status.
declare -A CNT # Count of consecutive status checks

while getopts ":c:" opt; do
  case ${opt} in
    c)
      CONFIG=$OPTARG
      ;;
    \? )
      >&2 echo "Invalid option: $OPTARG"
      exit 1
      ;;
    : )
      >&2 echo "Invalid option: $OPTARG requires an argument"
      exit 1
      ;;
  esac
done
shift $((OPTIND -1))

if [ ! -f "$CONFIG" ]; then
  >&2 echo "$CONFIG: no such file"
  exit 1
fi

# shellcheck source=/dev/null
. "$CONFIG"

for v in WAN_IF WAN_TABLE WAN_GW WAN_WEIGHT; do
  for i in 1 2; do
    var="${v}${i}"
    val="${!var}"

    if [ -n "$val" ]; then
      eval "${v}[$i]=$val"
    fi
  done
done

for i in "${!WAN_IF[@]}"; do
  WAN_IP[$i]="$(ip addr show "${WAN_IF[$i]}" 2> /dev/null | awk '$1 == "inet" { gsub(/\/.*/, "", $2); print $2 }')"

  if [ -z "${WAN_IF[$i]}" ] || [ -z "${WAN_TABLE[$i]}" ] || [ -z "${WAN_GW[$i]}" ] || [ -z "${WAN_IP[$i]}" ]; then
    unset "WAN_IF[$i]"
    continue
  fi

  LLS[$i]=1
  LPS[$i]=1
  CPS[$i]=1
  CLS[$i]=1
  CNT[$i]=0

  if [ -z "${WAN_WEIGHT[$i]}" ]; then
    WAN_WEIGHT[$i]=1
  fi
done

now() {
  date +"%a %b %d %T %Y"
}

log() {
  if [ -n "$LOG_FILE" ]; then
    echo -e "$(now): $__BASENAME: $*" | tee -a "$LOG_FILE" 1>&2
  else
    logger -p "$LOG_FACILITY.$LOG_PRIORITY" -t "${__BASENAME}[$$]" -s "$@"
  fi
}

link_status() {
  case $1 in
    0)
      echo "Up" ;;
    1)
      echo "Down" ;;
    *)
      echo "Unknown" ;;
  esac
}

# check_link $IP $TIMEOUT
check_link() {
  ping -W "$2" -I "$1" -c 1 "$PING_TARGET" > /dev/null 2>&1
  RETVAL=$?
  if [ $RETVAL -ne 0 ] ; then
    STATE=1
  else
    STATE=0
  fi

  link_status $STATE

  return $STATE
}

while : ; do
  for i in "${!WAN_IF[@]}"; do
    check_link "${WAN_IP[$i]}" "$PING_TIMEOUT"
    CPS[$i]=$?

    if [[ ${LPS[$i]} -ne ${CPS[$i]} ]] ; then
      log "Ping state changed for ${WAN_TABLE[$i]} from $(link_status ${LPS[$i]}) to $(link_status ${CPS[$i]})"
      CNT[$i]=1
    else
      if [[ ${LPS[$i]} -ne ${LLS[$i]} ]] ; then
        CNT[$i]=$((CNT[i]+1))
      fi
    fi

    if [[ ${CNT[$i]} -ge $SUCCESS_COUNT || (${LLS[$i]} -eq 0 && ${CNT[$i]} -ge $FAILURE_COUNT) ]]; then
      CLS[$i]=0
      CNT[$i]=0

      if [[ ${LLS[$i]} -eq 1 ]] ; then
        LLS[$i]=0
      else
        LLS[$i]=1
      fi
      log "Link state for ${WAN_TABLE[$i]} is $(link_status ${LLS[$i]})"
    else
      CLS[$i]=1
    fi

    LPS[$i]=${CPS[$i]}
  done

  for i in "${!CLS[@]}"; do
    if [[ ${CLS[$i]} -eq 0 ]]; then
      COMMAND=("ip route replace default scope global")

      if [[ $STARTUP -eq 1 ]] ; then
        STARTUP=0
        CHECK_INTERVAL=15
      fi

      for j in "${!LLS[@]}"; do
        if [[ ${LLS[$j]} -eq 0 ]]; then
          COMMAND+=("nexthop via ${WAN_GW[$j]} dev ${WAN_IF[$j]} weight ${WAN_WEIGHT[$j]}")
        fi
      done

      if [[ ${#COMMAND[@]} -gt 1 ]]; then
        eval "${COMMAND[@]}"

        if [ -n "$ON_CHANGE" ]; then
          log "Run 'ON_CHANGE' command"
          eval "$ON_CHANGE" 2>/dev/null || true
        fi
      fi

      break
    fi
  done

  sleep $CHECK_INTERVAL
done
